### DISCUSSION TITLE
Is there a way to find out if my application is running with the agent already active?

### DISCUSSION THREAD

**Discussion by evantorrie** [open-telemetry/opentelemetry-java-instrumentation#9173] [Q&A] [answered]
I have an OpenTelemetry aware application that up until now I've programmatically used the AutoConfigureSdk  and manual instrumentation via the OpenTelemetry API.

Some libraries have been introduced, and I'd now like to consider just starting the application up with the agent and turning off agent instrumentation for the libraries we have already instrumented manually.

Is there a way for my application to know that the agent is running?
More specifically, is there a way for me to either
a) pass the SDK I've configured in my app to the agent to use?
or
b) get a handle to the SDK that the agent instantiates via AutoConfigure?

**Comment by laurit** [comment, marked as answer]
> Is there a way for my application to know that the agent is running?

I think the easiest way is to `Class.forName("io.opentelemetry.javaagent.OpenTelemetryAgent")` or some other class that is present in the agent, and if it is found assume that you are running with the agent.

> a) pass the SDK I've configured in my app to the agent to use?

no

> b) get a handle to the SDK that the agent instantiates via AutoConfigure?

yes, use `GlobalOpenTelemetry.get()`

**Reply by evantorrie** [reply]
Thanks - this is awesome.

`GlobalOpenTelemetry.get()` returns an `OpenTelemetry` interface, but I presume (?) I can follow some of the [`unobfuscate()`](https://github.com/open-telemetry/opentelemetry-java/blob/288c5df620b6615799ae359951f0b48155bc5074/sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkTest.java#L70) code and try casting that to an `OpenTelemetrySDK` instance if I want to actually get access to the SDK itself?  Although I guess one question is whether the `io.opentelemetry.api.GlobalOpenTelemetry$ObfuscatedOpenTelemetry` classname is the same in the agent -- or whether it is shaded to something else?

Ideally, I'd also like to know if tracing is enabled, because for our continuing use of manual instrumentation - we have an optimization where we don't bother (for example) to add "tracing interceptors" if we know that tracing is disabled (currently we approximate that by checking whether `otel.sdk.disabled=true` or `otel.traces.exporter=none` in our AutoConfiguredOpenTelemetrySdkBuilder properties), although now I think about it - maybe just creating an arbitrary Tracer followed by a "startSpan" and check to see if it's returned as a "noop" span would mean less "peeking into internals".

**Reply by trask** [reply]
> if I want to actually get access to the SDK itself?

this can only be done in a [Java agent extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension)

**Reply by evantorrie** [reply]
I guess I can probably add myself as another one who would like to see something like https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/3261 possible

**Comment by evantorrie** [comment]
> > if I want to actually get access to the SDK itself?
> 
> this can only be done in a [Java agent extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension)

By implementing an `AgentListener` extension I presume??  

I have read over that section and seem to recall that Javaagent extensions are sandboxed and loaded with an isolated classLoader? 
 Since I don't have much working knowledge of classloaders and class accessibility, my question would be whether it's possible for a Javaagent extension to communicate info back to the application that's running under the Java agent?  e.g. could I capture the Autoconfigured SDK here, and make it available such that my application code could have access to it.
 
 And yes, I know that I shouldn't need to have access to the SDK anywhere else in my application - I think this is just trying to fit into some Otel legacy code where it was only possible to manually instantiate a logback OTLP appender by having access to the SDK to get a logEmitter.

Incidentally, I did try copy/pasting the [`unobfuscate` code](https://github.com/open-telemetry/opentelemetry-java/blob/288c5df620b6615799ae359951f0b48155bc5074/sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkTest.java#L70) into my application, but it fails at this point:

```java
            Object delegate = delegateField.get(openTelemetry);
```

with 

```
ERROR c.y.t.g.i.c.OtelConfiguration - Failed to get agent SDK
java.lang.IllegalArgumentException: Can not set final io.opentelemetry.api.OpenTelemetry field io.opentelemetry.api.GlobalOpenTelemetry$ObfuscatedOpenTelemetry.delegate to io.opentelemetry.javaagent.instrumentation.opentelemetryapi.v1_27.ApplicationOpenTelemetry127
        at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(Unknown Source)
        at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(Unknown Source)
        at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.ensureObj(Unknown Source)
        at java.base/jdk.internal.reflect.UnsafeQualifiedObjectFieldAccessorImpl.get(Unknown Source)
        at java.base/java.lang.reflect.Field.get(Unknown Source)
        at com.mycompany.instrumentation.config.AgentInterrogator.unobfuscate(AgentInterrogator.java:67)
```

which I guess answers my question as to whether the OpenTelemetry API is shaded.
